new_ret_no_self false positives

JoshMcguigan opened the pull request:
WORK IN PROGRESS

I plan to fix all of the false positives in #3313 in this PR, but I wanted to open it now to start gathering feedback.

In this first commit, I've updated the lint to allow tuple return types as long as `Self` shows up at least once, in any position of the tuple. I believe this is the broadest possible interpretation of what should be allowed for tuple return types, but I would certainly be okay making the lint more strict. 

fixes #3313

oli-obk commented on tests/ui/new_ret_no_self.rs:
```diff
@@ -91,3 +91,31 @@ impl V {
+
+impl TupleReturnerOk {
+    // should not trigger lint
+    pub fn new() -> (Self, u32) { unimplemented!(); }
```
`((Self, u32), u32)` Will also trigger the lint. I'm not sure at all how strict we should be. I mean it seems perfectly reasonable to allow `(Option<Self>, u32)` or at least `Option<(Self, u32)>`. This might get out of hand quickly.

If you want to cover a lot of cases, you can call [`walk`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/ty/struct.TyS.html#method.walk) on the return type (giving you an iterator) and iterate over everything.

That would still cause false positives for `struct Foo(pub Bar); impl Bar { fn new() -> Foo { ... } }` but we need to restrict ourselves somewhere I guess.

JoshMcguigan commented:
The suggestion to use `walk` cleaned this up significantly. But I still need to add some test cases to ensure `walk` covers all of our use cases.

JoshMcguigan commented:
I've added test cases which I believe cover all of the issues from #3313, and removed a clippy allow annotation within the clippy code base which should not be necessary anymore (we were returning `Option<Self>`.

Assuming we don't want to make the lint more strict I think this is good to go.

JoshMcguigan commented:
Any thoughts on why the travis build failed? I can't make out the reason looking at the logs.

flip1995 commented:
I restarted the job.

flip1995 commented:
For completeness (https://github.com/rust-lang-nursery/rust-clippy/pull/3338#discussion_r226625663) test cases for `((Self, u32), u32)` and `Option<(Self, u32)>` would be nice. After that this should be ready to merge.

JoshMcguigan commented:
@flip1995 Thanks for the feedback. I've added those test cases. Let me know if you see any additional issues.

flip1995 commented:
LGTM now!

bors r+

flip1995 commented:
bors r-

flip1995 commented:
It seems that bors also needs a `staging.tmp` and a `trying.tmp` branch.

bors r+
